man/man1/gfedquota.1
'\" t
.\"     Title: gfedquota
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 25 Aug 2015
.\"    Manual: Gfarm
.\"    Source: Gfarm
.\"  Language: English
.\"
.TH "GFEDQUOTA" "1" "25 Aug 2015" "Gfarm" "Gfarm"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
gfedquota \- edit file limits
.SH "SYNOPSIS"
.HP \w'\fBgfedquota\fR\ 'u
\fBgfedquota\fR [\fIoptions\fR]
.SH "DESCRIPTION"
.PP
\fBgfedquota\fR
is a editor of limits (quotas) for users and groups\&.
.PP
There are two kinds of limits which are "soft limit" and "hard limit"\&. Soft limits have a grace period\&. Types of limits are the total number of files, the total space of files, the total number of files include replicas and the total space of files include replicas\&. Setting \*(Aqdisable\*(Aq or \-1 indicates unlimited\&. If no option for limits is specified, no quota is imposed\&.
.PP
This command can be executed by an administrator in the gfarmadm group\&.
.SH "OPTIONS"
.PP
\fB\-P NAME, \-\-path=NAME\fR
.RS 4
Specifies a path name to specify a metadata server instead of the root metadata server\&.
.RE
.PP
\fB\-u NAME, \-\-user=NAME\fR
.RS 4
Specifies a user name to display the current usage and limits\&.
.RE
.PP
\fB\-g NAME, \-\-group=NAME\fR
.RS 4
Specifies a group name to display the current usage and limits\&.
.RE
.PP
\fB\-G SECOND, \-\-grace=SECOND\fR
.RS 4
Specifies a grace period in second for soft limits\&.
.RE
.PP
\fB\-s BYTE, \-\-softspc=BYTE\fR
.RS 4
Specifies a soft limit of file space in byte\&.
.RE
.PP
\fB\-h BYTE, \-\-hardspc=BYTE\fR
.RS 4
Specifies a hard limit of file space in byte\&.
.RE
.PP
\fB\-m NUM, \-\-softnum=NUM\fR
.RS 4
Specifies a soft limit of file number\&.
.RE
.PP
\fB\-n NUM, \-\-hardnum=NUM\fR
.RS 4
Specifies a hard limit of file number\&.
.RE
.PP
\fB\-S BYTE, \-\-physoftspc=BYTE\fR
.RS 4
Specifies a soft limit of physical space in byte include replicas\&.
.RE
.PP
\fB\-H BYTE, \-\-phyhardspc=BYTE\fR
.RS 4
Specifies a hard limit of physical space in byte include replicas\&.
.RE
.PP
\fB\-M NUM, \-\-physoftnum=NUM\fR
.RS 4
Specifies a soft limit of physical number include replicas\&.
.RE
.PP
\fB\-N NUM, \-\-phyhardnum=NUM\fR
.RS 4
Specifies a hard limit of physical number include replicas\&.
.RE
.PP
\fB\-?\fR
.RS 4
Displays a list of command options\&.
.RE
.SH "SEE ALSO"
.PP
\fBgfquota\fR(1),
\fBgfquotacheck\fR(1),
\fBgfusage\fR(1)


ChangeLog
2007-03-31  Noriyuki Soda  <soda@sra.co.jp>

	* RELNOTES
	avoid an expression which may be misinterpreted as a security problem.

2007-03-31  Osamu Tatebe <tatebe@cs.tsukuba.ac.jp>

	* RELNOTES
	correct description

2007-03-30  Tohru Sotoyama <sotoyama@sra.co.jp>

	* package/solaris/build
	change globus root directory

2007-03-30  Noriyuki Soda  <soda@sra.co.jp>

	* ChangeLog LICENSE RELNOTES
	merge changes in the CVS main trunk until gfarm-1.4.1 to the GNEO
	branch
	
	* ChangeLog
	Catched up with the current ChangeLog.

	* ChangeLog
	recover missed log messages in previous commit
	
	* doc/html/en/ref/man5/gfarm.conf.5.html, doc/html/ja/ref/man5/gfarm.conf.5.html man/ja/man5/gfarm.conf.5 man/man5/gfarm.conf.5:
	regen

	* doc/Gfarm-FAQ.en, doc/Gfarm-FAQ.ja, doc/docbook/en/ref/man5/gfarm.conf.5.docbook, doc/docbook/ja/ref/man5/gfarm.conf.5.docbook:
	merge the following changes in the SourceForge Subversion repository:

	Revision: 3674
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3674&view=rev
	Author:   sotoyama
	Date:     2007-03-29 02:34:39 -0700 (Thu, 29 Mar 2007)
	Log Message:
	add limitation about mixing gsi authentication with sharedsecret
	found in [gfarm-developers:02814]

	Revision: 3676
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3676&view=rev
	Author:   sotoyama
	Date:     2007-03-29 04:46:11 -0700 (Thu, 29 Mar 2007)
	Log Message:
	add limitation about mixing gsi authentication with sharedsecret
	found in [gfarm-developers:02814] to `auth' statement

	Revision: 3682
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3682&view=rev
	Author:   n-soda
	Date:     2007-03-29 11:03:31 -0700 (Thu, 29 Mar 2007)
	Log Message:
	swap section order of limitation and example of the "auth" directive,
	and add a paragraph to describe the example.

2007-03-30  yusuke Tanimura <yusuke.tanimura@aist.go.jp>

	* RELNOTES:
	proofreading

2007-03-29  Noriyuki Soda  <soda@sra.co.jp>

	* server/gfsd/gfsd.c:
	merge the follownig change from the SourceForge Subversion repository:

	Revision: 3681
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3681&view=rev
	Author:   n-soda
	Date:     2007-03-29 07:37:36 -0700 (Thu, 29 Mar 2007)
	Log Message:
	report error status of close(2) in gfs_server_open_local(),
	problem reported by Markus Elfring as:
	http://sourceforge.net/tracker/index.php?func=detail&aid=1687850&group_id=172718&atid=862742

	Also, treat an error at sendmsg(2) as just an error instead of
	fatal one.

	* bench/fsysbench/fsysbench.c:
	merge the follownig change from the SourceForge Subversion repository:

	Revision: 3679
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3679&view=rev
	Author:   n-soda
	Date:     2007-03-29 05:13:05 -0700 (Thu, 29 Mar 2007)
	Log Message:
	timer_interrupt_handler() shouldn't call printf(3), but write(2),
	because the former is not async-signal-safe.
	c.f.
	https://sourceforge.net/tracker/?func=detail&atid=862742&aid=1687842&group_id=172718
	
	* ftool/gfarm_agent/gfarm_agent.c:
	merge the follownig change from the SourceForge Subversion repository:

	Revision: 3678
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3678&view=rev
	Author:   n-soda
	Date:     2007-03-29 04:59:54 -0700 (Thu, 29 Mar 2007)
	Log Message:
	sigterm_handler() shouldn't call exit(3), but _exit(2), because
	the former is not async-signal-safe.
	c.f.
	https://sourceforge.net/tracker/?func=detail&atid=862742&aid=1687842&group_id=172718
	
	* server/gfsd/gfsd.c:
	merge the follownig change from the SourceForge Subversion repository:

	Revision: 3675
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3675&view=rev
	Author:   n-soda
	Date:     2007-03-29 03:30:02 -0700 (Thu, 29 Mar 2007)
	Log Message:
	do not call exit(3) but _exit(2) in accepting_sigterm_handler(),
	because the former is not async-signal-safe.
	problem reported by Markus Elfring in
	http://sourceforge.net/tracker/index.php?func=detail&aid=1687842&group_id=172718&atid=862742

2007-03-29  yusuke Tanimura <yusuke.tanimura@aist.go.jp>

	* RELNOTES, LICENSE:
	update release notes and the copyright display for the version 1.4.1

2007-03-28  Osamu Tatebe <tatebe@cs.tsukuba.ac.jp>

	* ChangeLog:
	Catched up with the current ChangeLog.

2007-03-28  Noriyuki Soda  <soda@sra.co.jp>

	* include/gfarm/gfarm_misc.h, package/freebsd/Makefile,	package/pkgsrc/Makefile, package/redhat/gfarm.spec, package/solaris/build:
	increase version number to 1.4.1

	* include/gfarm/gfarm_error.h, lib/libgfarm/gfarm/error.c, lib/libgfarm/gfarm/gfs_client.c:
	merge the following change from the SourceForge Subversion repository:

	Revision: 3673
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3673&view=rev
	Author:   n-soda
	Date:     2007-03-28 07:11:37 -0700 (Wed, 28 Mar 2007)
	Log Message:
	make gfsd reconnection feature work against more network errors.
	this fixes a problem reported by Yamamoto Naotaka-san
	in [gfarm-developers:03456] which was caused by ECONNRESET.

	introduce ENETDOWN, EHOSTDOWN and ENETRESET as gfarm native errors
	for this purpose. (although ENETRESET will be nearly never used)

	* gftool/gneo/gneo.c:
	just a portability fix

	* regress/gfs_hook/lib/fsystest/index.FreeBSD.expected, regress/gfs_hook/lib/fsystest/index.Linux.expected, regress/gfs_hook/lib/fsystest/index.NetBSD.expected, regress/gfs_hook/lib/fsystest/index.SunOS.expected:
	results are different on the GNEO branch with gfs_hook


gftool/gfquota/gfedquota/gfedquota.c
/*
 * $Id$
 */

#include <stdio.h>
#include <stdlib.h>
#include <libgen.h>
#include <unistd.h>
#include <string.h>

#include <gfarm/gfarm.h>

#ifdef HAVE_GETOPT_LONG
#include <getopt.h>
#endif

#include "config.h"
#include "quota_info.h"
#include "gfm_client.h"
#include "lookup.h"
#include "gfarm_path.h"

const char *program_name = "gfedquota";

#ifdef HAVE_GETOPT_LONG
#define OPT(s, l, m) fprintf(stderr, "  -%s,--%s\t%s\n", s, l, m)
#else
#define OPT(s, l, m) fprintf(stderr, "  -%s\t%s\n", s, m)
#endif

static void
usage(void)
{
	fprintf(stderr,
		"Usage:"
		"\t%s -u <user_name> [options]\n"
		"\t%s -g <group_name>) [options]\n"
		"\t%s -D <dirset_name> [-u <user_name>] [options]\n"
		"Options:\n"
		"(If the value is 'disable' or -1, the limit is disabled):\n",
		program_name, program_name, program_name);
	OPT("P", "path=NAME", "pathname "
		"(to select one of multiple metadata servers)");
	OPT("u", "user=NAME", "username");
	OPT("g", "group=NAME", "groupname");
	OPT("D", "dirset=NAME", "dirset name (for directory quota)");
	OPT("G", "grace=SEC", "grace period for all SoftLimits");
	OPT("s", "softspc=BYTE", "SoftLimit of total used file space");
	OPT("h", "hardspc=BYTE", "HardLimit of total used file space");
	OPT("m", "softnum=NUM", "SoftLimit of total used file number");
	OPT("n", "hardnum=NUM", "HardLimit of total used file number");
	OPT("S", "physoftspc=BYTE", "SoftLimit of total used physical space");
	OPT("H", "phyhardspc=BYTE", "HardLimit of total used physical space");
	OPT("M", "physoftnum=NUM", "SoftLimit of total used physical number");
	OPT("N", "phyhardnum=NUM", "HardLimit of total used physical number");
	OPT("?", "help", "this help message");

	exit(2);
}

static void
gfarm_quota_set_info_to_limit_info(
	const struct gfarm_quota_set_info *si,
	struct gfarm_quota_limit_info *li)
{
	li->grace_period = si->grace_period;
	li->soft.space = si->space_soft;
	li->hard.space = si->space_hard;
	li->soft.num = si->num_soft;
	li->hard.num = si->num_hard;
	li->soft.phy_space = si->phy_space_soft;
	li->hard.phy_space = si->phy_space_hard;
	li->soft.phy_num = si->phy_num_soft;
	li->hard.phy_num = si->phy_num_hard;
}

static gfarm_int64_t
convert_value(const char *str)
{
	gfarm_int64_t val;
	char *endptr;

	if (strcmp(str, "disable") == 0)
		return (GFARM_QUOTA_INVALID);

	val = strtoll(str, &endptr, 10);
	if (*endptr != '\0') {
		fprintf(stderr, "ignore invalid parameter: %s\n", str);
		return (GFARM_QUOTA_NOT_UPDATE);
	}
	if (val == -1)
		return (GFARM_QUOTA_INVALID);
	else if (val < -1)
		return (GFARM_QUOTA_NOT_UPDATE);
	else
		return (val);
}

int
main(int argc, char **argv)
{
	gfarm_error_t e;
	int c, status = 0;
	char *username = NULL, *groupname = NULL, *dirsetname = NULL;
	char *optstring = "P:u:D:g:G:s:h:m:n:S:H:M:N:?";
#ifdef HAVE_GETOPT_LONG
	struct option long_options[] = {
		{"path", 1, NULL, 'P'},
		{"user", 1, NULL, 'u'},
		{"dirset", 1, NULL, 'D'},
		{"group", 1, NULL, 'g'},
		{"grace", 1, NULL, 'G'},
		{"softspc", 1, NULL, 's'},
		{"hardspc", 1, NULL, 'h'},
		{"softnum", 1, NULL, 'm'},
		{"hardnum", 1, NULL, 'n'},
		{"physoftspc", 1, NULL, 'S'},
		{"phyhardspc", 1, NULL, 'H'},
		{"physoftnum", 1, NULL, 'M'},
		{"phyhardnum", 1, NULL, 'N'},
		{"help", 0, NULL, '?'},
		{0, 0, 0, 0}
	};
#endif
	struct gfarm_quota_set_info qi = {
		NULL,
		GFARM_QUOTA_NOT_UPDATE,
		GFARM_QUOTA_NOT_UPDATE,
		GFARM_QUOTA_NOT_UPDATE,
		GFARM_QUOTA_NOT_UPDATE,
		GFARM_QUOTA_NOT_UPDATE,
		GFARM_QUOTA_NOT_UPDATE,
		GFARM_QUOTA_NOT_UPDATE,
		GFARM_QUOTA_NOT_UPDATE,
		GFARM_QUOTA_NOT_UPDATE,
	};
	struct gfm_connection *gfm_server;
	const char *path = ".";
	char *realpath = NULL;

	if (argc > 0)
		program_name = basename(argv[0]);
	e = gfarm_initialize(&argc, &argv);
	if (e != GFARM_ERR_NO_ERROR) {
		fprintf(stderr, "%s: %s\n", program_name,
			gfarm_error_string(e));
		exit(1);
	}

	while (1) {
#ifdef HAVE_GETOPT_LONG
		int option_index = 0;
		c = getopt_long(argc, argv, optstring,
				long_options, &option_index);
#else
		c = getopt(argc, argv, optstring);
#endif
		if (c == -1)
			break;
		switch (c) {
		case 'P':
			path = optarg;
			break;
		case 'u':
			username = optarg;
			break;
		case 'g':
			groupname = optarg;
			break;
		case 'D':
			dirsetname = optarg;
			break;
		case 'G':
			qi.grace_period = convert_value(optarg);
			break;
		case 's':
			qi.space_soft = convert_value(optarg);
			break;
		case 'h':
			qi.space_hard = convert_value(optarg);
			break;
		case 'm':
			qi.num_soft = convert_value(optarg);
			break;
		case 'n':
			qi.num_hard = convert_value(optarg);
			break;
		case 'S':
			qi.phy_space_soft = convert_value(optarg);
			break;
		case 'H':
			qi.phy_space_hard = convert_value(optarg);
			break;
		case 'M':
			qi.phy_num_soft = convert_value(optarg);
			break;
		case 'N':
			qi.phy_num_hard = convert_value(optarg);
			break;
		case '?':
		default:
			usage();
		}
	}
	argc -= optind;
	argv += optind;

	if (gfarm_realpath_by_gfarm2fs(path, &realpath) == GFARM_ERR_NO_ERROR)
		path = realpath;
	if ((e = gfm_client_connection_and_process_acquire_by_path(
		     path, &gfm_server)) != GFARM_ERR_NO_ERROR) {
		fprintf(stderr, "%s: metadata server for \"%s\": %s\n",
			program_name, path, gfarm_error_string(e));
		status = 1;
		goto terminate;
	}
	free(realpath);

	if (username != NULL && groupname == NULL && dirsetname == NULL) {
		qi.name = username;
		e = gfm_client_quota_user_set(gfm_server, &qi);
	} else if (groupname != NULL &&
	    username == NULL && dirsetname == NULL) {
		qi.name = groupname;
		e = gfm_client_quota_group_set(gfm_server, &qi);
	} else if (dirsetname != NULL && groupname == NULL) {
		struct gfarm_quota_limit_info li;

		gfarm_quota_set_info_to_limit_info(&qi, &li);
		e = gfm_client_quota_dirset_set(gfm_server,
		    username != NULL ? username :
		    gfm_client_username(gfm_server),
		    dirsetname, &li);
	} else
		usage();
	if (e != GFARM_ERR_NO_ERROR) {
		fprintf(stderr, "%s: %s\n",
		    program_name, gfarm_error_string(e));
		status = 1;
	}
	gfm_client_connection_free(gfm_server);
terminate:
	e = gfarm_terminate();
	if (e != GFARM_ERR_NO_ERROR) {
		fprintf(stderr, "%s: %s\n", program_name,
		    gfarm_error_string(e));
		status = 1;
	}
	return (status);
}
